### PR TITLE
[Merged by Bors] - feat(RingTheory/Invariant): Define invariant extensions of rings

### PR DESCRIPTION
**Pull request by tb65536**
This PR adds a typeclass for `∀ b : B, (∀ g : G, g • b = b) → ∃ a : A, algebraMap A B a = b`, working towards Frobenius elements in #17717.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Invariant.lean [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+-/
+
+namespace Algebra
+
+variable (A B G : Type*) [CommSemiring A] [Semiring B] [Algebra A B]
+  [Group G] [MulSemiringAction G B]
+
+/-- An action of a group `G` on an extension of rings `B/A` is invariant if every fixed  -/
+class IsInvariant : Prop where
+  isInvariant : ∀ b : B, (∀ g : G, g • b = b) → ∃ a : A, algebraMap A B a = b
```
Comment (erdOne) [inline Mathlib/RingTheory/Invariant.lean:29]:
```suggestion
/-- An action of a group `G` on an extension of rings `B/A` is invariant if every fixed  -/
@[mk_iff]
class IsInvariant : Prop where
  isInvariant : ∀ b : B, (∀ g : G, g • b = b) → ∃ a : A, algebraMap A B a = b
```

`mk_iff` generates the `isInvariant_def` lemma below.
Also the docstring doesn't seem finished.

I wonder if stating it as `FixedPoints.subsemiring G B ≤ (algebraMap A B).srange` would be useful but you'll have to define `FixedPoints.subsemiring` first, so maybe not.

And also I'm surprised you don't need equality here.

File: Mathlib/RingTheory/Invariant.lean [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  rw [charpoly_eq, eval_prod]
+  apply Finset.prod_eq_zero (Finset.mem_univ (1 : G))
+  rw [one_smul, eval_sub, eval_C, eval_X, sub_self]
+
+variable {G}
+
+theorem charpoly_smul (b : B) (g : G) : g • (charpoly G b) = charpoly G b := by
+  rw [charpoly_eq_prod_smul, Finset.smul_prod_perm]
+
+private theorem charpoly_coeff_smul (b : B) (n : ℕ) (g : G) :
```
Comment (erdOne) [inline Mathlib/RingTheory/Invariant.lean:70]: Why is this private?

File: Mathlib/RingTheory/Invariant.lean [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+variable (A B G : Type*) [CommRing A] [CommRing B] [Algebra A B] [Group G] [MulSemiringAction G B]
+
+namespace MulSemiringAction
+
+open Polynomial
+
+variable {B} [Fintype G]
+
+/-- Characteristic polynomial of a group action on a ring. -/
```
Comment (erdOne) [inline Mathlib/RingTheory/Invariant.lean:49]:
```suggestion
/-- Characteristic polynomial of a finite group action on a ring. -/
```

File: Mathlib/RingTheory/Invariant.lean (hunk 2 above)
Comment (tb65536) [reply Mathlib/RingTheory/Invariant.lean:70]: Oops, thanks!

File: Mathlib/RingTheory/Invariant.lean (hunk 1 above)
Comment (tb65536) [reply Mathlib/RingTheory/Invariant.lean]: Equality can be achieved by adding a `SMulCommClass` assumption. I'll update the docstring.

File: Mathlib/RingTheory/Invariant.lean [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+end MulSemiringAction
+
+namespace Algebra.IsInvariant
+
+open MulSemiringAction Polynomial
+
+variable [IsInvariant A B G]
+
+theorem exists_map_eq_charpoly [Fintype G] (b : B) :
```
Comment (riccardobrasca) [inline Mathlib/RingTheory/Invariant.lean:78]: Can you use [Polynomial.lifts](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Polynomial/Lifts.html#Polynomial.lifts) here? It has a nice API, and if something is missing it should be added directly there.

Comment (tb65536) [reply Mathlib/RingTheory/Invariant.lean]: Nice, thanks!

File: Mathlib/RingTheory/Invariant.lean [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+/-!
+# Invariant Extensions of Rings
+
+Given an extension of rings `B/A` and an action of `G` on `B`, we introduce a predicate
+`Algebra.IsInvariant A B G` which states that every fixed point of `B` lies in the image of `A`.
+
+## Main statements
+* `Algebra.IsInvariant.isIntegral`: If `G` is finite, then `B/A` is an integral extension.
+
+TODO: Prove the existence of Frobenius elements in this general setting.
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/Invariant.lean:19]: Please point at your pr for this

File: Mathlib/RingTheory/Invariant.lean [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  rw [charpoly_eq, eval_prod]
+  apply Finset.prod_eq_zero (Finset.mem_univ (1 : G))
+  rw [one_smul, eval_sub, eval_C, eval_X, sub_self]
+
+variable {G}
+
+theorem charpoly_smul (b : B) (g : G) : g • (charpoly G b) = charpoly G b := by
+  rw [charpoly_eq_prod_smul, Finset.smul_prod_perm]
+
+theorem charpoly_coeff_smul (b : B) (n : ℕ) (g : G) :
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/Invariant.lean:67]: Is this not `smul_coeff`?

File: Mathlib/RingTheory/Invariant.lean [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  monic_prod_of_monic _ _ (fun _ _ ↦ monic_X_sub_C _)
+
+theorem charpoly_eval (b : B) : (charpoly G b).eval b = 0 := by
+  rw [charpoly_eq, eval_prod]
+  apply Finset.prod_eq_zero (Finset.mem_univ (1 : G))
+  rw [one_smul, eval_sub, eval_C, eval_X, sub_self]
+
+variable {G}
+
+theorem charpoly_smul (b : B) (g : G) : g • (charpoly G b) = charpoly G b := by
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/Invariant.lean:64]: `smul_charpoly`

**Review by Ruben-VandeVelde — APPROVED** [review]
Lgtm modulo names

**Comment by kbuzzard** [comment]
Thanks a lot!

bors merge

**Pull request closed without merging**